cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,857 @@
+};
+
+//---------------------------------------------------------------------------//
+// Hierarchical index spaces
+// SparseIndexSpace <- BlockIndexSpace <- TileIndexSpace
+// Naming:
+//      Block = MPI Rank
+//      Tile = Sub-block with several cells
+//      Cell = Basic grid unit
+//---------------------------------------------------------------------------//
+
+//! Declaration of BlockIndexSpace
+template <typename MemorySpace, unsigned long long CBits,
+          unsigned long long CNumPerDim, unsigned long long CNumPerTile,
+          HashTypes Hash, typename Key, typename Value>
+class BlockIndexSpace;
+
+//! Declaration of TileIndexSpace
+template <int CBits, int CNumPerDim, int CNumPerTile>
+class TileIndexSpace;
+
+/*!
+  \class SparseIndexSpace
+  \brief Sparse index space, with a hierarchical structure (cell->tile->block)
+  \tparam MemorySpace Memory space to store the IndexSpace(Hash Table)
+  \tparam CellPerTileDim Cell number inside each tile per dimension
+  \tparam Hash Hash type (lexicographical or morton)
+  \tparam Key Type of the tile/cell hash key
+  \tparam Value Type of the tile/cell No.
+ */
+template <typename MemorySpace, unsigned long long CellPerTileDim = 4,
+          HashTypes Hash = HashTypes::Naive, typename Key = uint64_t,
+          typename Value = uint64_t>
+class SparseIndexSpace
+{
+  public:
+    //! Number of dimensions, 3 = ijk
+    static constexpr int rank = 3;
+    //! Number of bits (per dimension) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile_dim =
+        bitCount( CellPerTileDim );
+    //! Number of cells inside each tile (per dimension), tile size reset to
+    //! power of 2
+    static constexpr unsigned long long cell_num_per_tile_dim =
+        1 << cell_bits_per_tile_dim;
+    //! Cell mask (per dimension), indicating which part of the tile
+    //! address(binary bits) will index the cells inside each tile
+    static constexpr unsigned long long cell_mask_per_tile_dim =
+        ( 1 << cell_bits_per_tile_dim ) - 1;
+    //! Number of bits (total) needed to index the cells inside a tile
+    static constexpr unsigned long long cell_bits_per_tile =
+        cell_bits_per_tile_dim + cell_bits_per_tile_dim +
+        cell_bits_per_tile_dim;
+    //! Number of cells (total) inside each tile
+    static constexpr unsigned long long cell_num_per_tile =
+        cell_num_per_tile_dim * cell_num_per_tile_dim * cell_num_per_tile_dim;
+    //! Types
+    using key_type = Key;                        // tile hash key
+    using value_type = Value;                    // tile No.
+    static constexpr HashTypes hash_type = Hash; // hash table
+
+    /*!
+      \brief Constructor
+      \param size The size of the block (MPI rank) (Unit: cell)
+      \param capacity Expected capacity of the allocator to store the tiles
+      \param rehash_factor The factor multiplies to the capacity for resizing,
+      when tile nums exceed the capcity
```

Comment:
Yes, 0 should work. So let me make it clearer for the capacity(I'm not sure if it's the best design). We actually have two capacities — one for the unordered map, another for the pre-allocation size.

1. For the unordered map, I set the default size as the number of all uniform grids (`size_x * size_y * size_z`). The reason is that the insertion is executed on the device in a parallel way. During this process, it seems that Kokko unordered map cannot automatically resize. So I have to make sure that the inserting number does not exceed the hash table's capacity. If I do a full insertion (insert all uniform grids) in parallel with a smaller (unordered map) capacity, I would have a runtime error. Also, it would worth noticing that we cannot exactly control the capacity of Kokkos unordered map. The capacity value we feed to the unordered map would finally be a lower bound instead of an exact size. 

2. For the pre-allocation size, I use `capacity` and `rehash_factor` here for allocation resize purposes. (Well, when writing to this point, I think it would be better to rename these two variables to `prealloc_size` and `realloc_factor`). For this part, I only did the initialization operations. I plan to leave the resizes operations to the allocation PR to make things more consistent. Or maybe we can consider moving these two parameters allocation related structs? Still thinking about this design.